extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaSimpleConsumer.java

```diff
@@ -0,0 +1,405 @@
+    requestInfo.put(
+        topicAndPartition,
+        new PartitionOffsetRequestInfo(
+            earliest ? kafka.api.OffsetRequest.EarliestTime() : kafka.api.OffsetRequest.LatestTime(), 1
+        )
+    );
+    OffsetRequest request = new OffsetRequest(
+        requestInfo, kafka.api.OffsetRequest.CurrentVersion(), clientId
+    );
+    OffsetResponse response = null;
+    try {
+      response = consumer.getOffsetsBefore(request);
+    }
+    catch (Exception e) {
+      ensureNotInterrupted(e);
+      log.error(e, "caught exception in getOffsetsBefore [%s] - [%s]", topic, partitionId);
+      return -1;
+    }
+    if (response.hasError()) {
+      log.error(
+          "error fetching data Offset from the Broker [%s]. reason: [%s]", leaderBroker.host(),
+          response.errorCode(topic, partitionId)
+      );
+      return -1;
+    }
+    long[] offsets = response.offsets(topic, partitionId);
+    return earliest ? offsets[0] : offsets[offsets.length - 1];
+  }
+
+  public Iterable<BytesMessageWithOffset> fetch(long offset, int timeoutMs) throws InterruptedException
+  {
+    FetchResponse response = null;
+    Broker previousLeader = leaderBroker;
+    while (true) {
+      ensureConsumer(previousLeader);
+
+      FetchRequest request = new FetchRequestBuilder()
+          .clientId(clientId)
+          .addFetch(topic, partitionId, offset, FETCH_SIZE)
+          .maxWait(timeoutMs)
+          .minBytes(1)
+          .build();
+
+      log.debug("fetch offset %s", offset);
+
+      try {
+        response = consumer.fetch(request);
+      }
+      catch (Exception e) {
+        ensureNotInterrupted(e);
+        log.warn(e, "caughte exception in fetch {} - {}", topic, partitionId);
+        response = null;
+      }
+
+      if (response == null || response.hasError()) {
+        short errorCode = response != null ? response.errorCode(topic, partitionId) : ErrorMapping.UnknownCode();
+        log.warn("fetch %s - %s with offset %s encounters error: [%s]", topic, partitionId, offset, errorCode);
+
+        boolean needNewLeader = false;
+        if (errorCode == ErrorMapping.RequestTimedOutCode()) {
+          log.info("kafka request timed out, response[%s]", response);
+        } else if (errorCode == ErrorMapping.OffsetOutOfRangeCode()) {
+          long newOffset = getOffset(earliest);
+          log.info("got [%s] offset[%s] for [%s][%s]", earliest ? "earliest" : "latest", newOffset, topic, partitionId);
+          if (newOffset < 0) {
+            needNewLeader = true;
+          } else {
+            offset = newOffset;
+            continue;
+          }
+        } else {
+          needNewLeader = true;
+        }
+
+        if (needNewLeader) {
+          stopConsumer();
+          previousLeader = leaderBroker;
+          leaderBroker = null;
+          continue;
+        }
+      } else {
+        break;
```

Comment:
agree with the logic being complex, however since this code is being used by the author in a production setup and does not really have unit tests, I am kind of uncomfortable changing it at this stage and breaking it.
I will probably be getting back to this code in some time (as part of kafka ingestion overhaul) and will try to improve it then.